### PR TITLE
🎉TESTING 🎉

### PR DESCRIPTION
**Pull request by gregkohn**
We've had some bugs go out, and the more repos we have pointing at this the more rock-solid we need it to be. 

Trevor found [this walkthrough](https://www.oliverdavies.uk/articles/testing-tailwindcss-plugins-with-jest/) which was fairly helpful and explained some concepts to me, but ultimately the author's Github account had what I was looking for.

I based this off of two things he built:
- [Testing helpers](https://github.com/opdavies/tailwindcss-plugin-test-helpers)
- [Example plugin with test](https://github.com/opdavies/tailwindcss-plugin-outline-offset/blob/master/test.js)

I started out by just installing his helpers, but ran into two limitations that forced me to bring them in-house and tweak:
1. They were not updated for the new `plugin()` API which we are using
1. They always merged the user config options defined in your test into the default Tailwind config. I've made this optional based on a param because it meant for something like our variant plugins, which need to be tested using a first-party plugin like `rotate`, you would have to generate **all** of the classes for the default `rotate` options (in that case, `0`, `45`, `-45`, `90`, `-90`, `180`, `-180`)

For all of our plugins that are not variants themselves, I included a variant to test that output as well. That seems like a good requirement to me. I do feel like it might be _more correct_ to make those a second test `it generates the sr classes` **and** `it generates the sr class variants`, but this meant a ton of duplication. Curious what others think though.

### CODE REVIEW

File: plugins/animation/test.js [hunk 1]
```diff
@@ -0,0 +1,66 @@
...
+          },
+        },
+      },
+    },
+  }
+
+  const output = `
+    .animation-spin {
+      animation: spin 1s infinite linear;
+    }
```
Comment (davist11) [inline plugins/animation/test.js:34]: Just curious do spaces/indentation matter?

File: plugins/blend/test.js [hunk 2]
```diff
@@ -0,0 +1,51 @@
...
+    }
+    .hover\\:bg-blend-multiply:hover {
+      background-blend-mode: multiply
+    }
+    .hover\\:blend-screen:hover {
+      mix-blend-mode: screen
+    }
+    .hover\\:bg-blend-screen:hover {
+      background-blend-mode: screen
+    }
```
Comment (davist11) [inline plugins/blend/test.js:43]: should we standardize this plugin to use semicolons?

**Review by davist11 — APPROVED** [review]
This is awesome 👏  nice work. Now just get that thing connected to Circle 😜

File: plugins/animation/test.js (hunk 1 above)
Comment (gregkohn) [reply plugins/animation/test.js:34]: It does not, neither do semicolons, which is why I missed them in your other comment.
It uses this: https://github.com/benface/jest-matcher-css/blob/master/index.js
Which, funnily enough, was written by the creator of Tailwind

File: .nvmrc [hunk 3]
```diff
@@ -0,0 +1 @@
+v12.10.0
```
Comment (leobauza) [inline .nvmrc:1]: any chance we can add a `.tool-versions` file if there isn't one already?

**Review by mzlock — APPROVED** [review]
👏

File: plugins/animation/test.js [hunk 4]
```diff
@@ -0,0 +1,66 @@
...
+      }
+    }
+  `
+
+  expect.assertions(2)
+  return generateUtilities(plugin, config).then((result) => {
+    expect(result.warnings().length).toBe(0)
+    expect(result.css).toMatchCss(output)
+  })
+})
```
Comment (leobauza) [inline plugins/animation/test.js:66]: this feels like what "snapshots" are made for but...the gap between me knowing this seems like what "snapshots" are made of and giving you instructions for that is vast...we can look at the Kent C. Dodds testing thing on snapshots tho?

**Review by leobauza — APPROVED** [review]
😎

**Review by jeremyfrank — APPROVED** [review]
Cool stuff! I pulled it down and it's all working.

**Pull request merged**